src/main/java/org/apache/solr/benchmarks/indexing/IndexBatchSupplier.java

```diff
@@ -0,0 +1,130 @@
+
+    private AtomicLong docsIndexed = new AtomicLong();
+
+    public IndexBatchSupplier(DocReader docReader, IndexBenchmark benchmark, DocCollection docCollection, HttpClient httpClient, Map<String, String> shardVsLeader) {
+        this.benchmark = benchmark;
+        this.docCollection = docCollection;
+
+        this.httpClient = httpClient;
+        this.shardVsLeader = shardVsLeader;
+
+        this.workerFuture = startWorker(docReader);
+    }
+
+    class IdParser implements JsonRecordReader.Handler {
+        String idParsed;
+
+        @Override
+        public void handle(Map<String, Object> record, String path) {
+            idParsed = record.get(benchmark.idField) instanceof String ?
+                    (String) record.get(benchmark.idField) :
+                    record.get(benchmark.idField).toString();
+        }
+    }
+
+    private Future<?> startWorker(DocReader docReader) {
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        //a continuously running job until either all input is exhausted or exit is flagged
+        Future<?> workerFuture = executorService.submit(() -> {
+            List<String> inputDocs;
+            JsonRecordReader rdr = JsonRecordReader.getInst("/", Collections.singletonList(benchmark.idField + ":/" + benchmark.idField));
+            IdParser idParser = new IdParser();
+            Map<String, List<String>> shardVsDocs = new HashMap<>();
+            try {
+                while (!exit && (inputDocs = docReader.readDocs(benchmark.batchSize)) != null) { //can read more than batch size, just use batch size as a sensible value
+                    for (String inputDoc : inputDocs) {
+                        rdr.streamRecords(new StringReader(inputDoc), idParser);
+                        Slice targetSlice = docCollection.getRouter().getTargetSlice(idParser.idParsed, null, null, null, docCollection);
+                        List<String> shardDocs = shardVsDocs.get(targetSlice.getName());
+                        if (shardDocs == null) {
+                            shardVsDocs.put(targetSlice.getName(), shardDocs = new ArrayList<>(benchmark.batchSize));
+                        }
+                        shardDocs.add(inputDoc);
+                        if (shardDocs.size() >= benchmark.batchSize) {
+                            shardVsDocs.remove(targetSlice.getName());
+
+                            //a shard has accumulated enough docs to be executed
+                            UploadDocs uploadDocs = new UploadDocs(shardDocs, httpClient, shardVsLeader.get(targetSlice.getName()), docsIndexed);
+                            while (!exit && !pendingBatches.offer(uploadDocs, 1, TimeUnit.SECONDS)) {
+                                //try again
+                            }
+                        }
+                    }
+                }
+                shardVsDocs.forEach((shard, docs) -> { //flush the remaining ones
```

Comment:
so, accordig to this logic , if `shard1` has crossed the batch size, all shards are uploaded? is this the intent?

Reply:
This is to mimic the old behavior at https://github.com/fullstorydev/solr-bench/pull/50/files/07565fdeb86670e2e32057cb893d0605a836a14e#diff-c443b311a116a519dcdbf35939db1b752571af41e44a38f37368dc4770092ab1L358

which after all lines are read from the source, it would flush out whatever is left that are added to shardVsDocs map, yet didn't pushed out as it has not reached the batch size. Take note that the logic here would only send out the rest if `exit` is not true (ie noone has called `close()` on this supplier yet). Therefore we can break in down into below scenarios:
1. If noone has explicitly called `close`, the reader would just read all the lines and this code in `IndexBatchSupplier` would flush the remaining docs in `shardVsDocs`
2. If someone has called `close`,  `IndexBatchSupplier` should not attempt to queue any more pending batches. Take note that no current code would do this while the supplier is active. This is just to put into place for the completeness of implementation (and probably some future usage can make use of this)


Take note that this supplier is unaware of `duration/maxExecution` or `maxDocs` settings for the benchmark. Such 2 restrictions are handled elsewhere:
1. `duration/maxExecution` by `ControlledExecuctor`, such executor can decide to drop the remaining task (if duration elapsed) or continue with the rest of the submitted batch (if submitted count reaches `maxExecution`)
2. `maxDocs` is handled by the `DocReader`, once it reaches the `maxDocs`, `readDocs` method should return `null`